utilities/blob_db/blob_db_impl.cc

```diff
@@ -462,6 +480,83 @@ void BlobDBImpl::ProcessCompactionJobInfo(const CompactionJobInfo& info) {
       LinkSstToBlobFile(output.file_number, output.oldest_blob_file_number);
     }
   }
+
+  MarkUnreferencedBlobFilesObsolete();
+}
+
+bool BlobDBImpl::MarkBlobFileObsoleteIfNeeded(
+    const std::shared_ptr<BlobFile>& blob_file, SequenceNumber obsolete_seq) {
+  assert(blob_file);
+  assert(!blob_file->HasTTL());
+  assert(blob_file->Immutable());
+  assert(bdb_options_.enable_garbage_collection);
+
+  // Note: FIFO eviction could have marked this file obsolete already.
+  if (blob_file->Obsolete()) {
+    return true;
+  }
+
+  if (!blob_file->GetLinkedSstFiles().empty()) {
+    return false;
+  }
+
+  ROCKS_LOG_INFO(db_options_.info_log,
+                 "Blob file %" PRIu64 " is no longer needed, marking obsolete",
+                 blob_file->BlobFileNumber());
+
+  ObsoleteBlobFile(blob_file, obsolete_seq, /* update_size */ true);
+  return true;
+}
+
+template <class Functor>
+void BlobDBImpl::MarkUnreferencedBlobFilesObsoleteImpl(Functor mark_if_needed) {
+  assert(bdb_options_.enable_garbage_collection);
+
+  // Iterate through all live immutable non-TTL blob files, and mark them
+  // obsolete assuming no SST files rely on the blobs in them.
+  // Note: we need to stop as soon as we find a blob file that has any
+  // linked SSTs.
```

Comment:
I'm not sure I quite get it. It assumes that obsolete files will happen in number order, so that once we see a file that is not obsolete, other files cannot be obsolete?

Reply:
In a nutshell, yes. As an example, let's assume the following mapping: SST #1 is linked to blob file #5, SST #2 is linked to blob file #7, and SST #3 is linked to blob file #4. A link here means that the blob file is the *oldest* (lowest-numbered) blob file the SST relies on. In other words, what we now based on this mapping is that SST #1 relies on blob files with numbers #5 *and up*, SST #2 on blob files #7 *and up*, and SST #3 on blob files #4 *and up*. Based on this, the blob files that are no longer needed (and can be marked obsolete) are blob files #1 through #3.